### PR TITLE
Protect the resource during access from being released

### PR DESCRIPTION
**Pull request by catostrophe**
Rework of #9

### CODE REVIEW

File: modules/core/src/main/scala/io/janstenpickle/hotswapref/HotswapRef.scala
Comment (janstenpickle) [inline modules/core/src/main/scala/io/janstenpickle/hotswapref/HotswapRef.scala:51]: I love this new implementation, but it could probably do with comment

File: modules/core/src/test/scala/io/janstenpickle/hotswapref/HotswapRefSpec.scala
Comment (catostrophe) [reply modules/core/src/test/scala/io/janstenpickle/hotswapref/HotswapRefSpec.scala]: removed them

**Review by janstenpickle — APPROVED** [review]

File: modules/core/src/main/scala/io/janstenpickle/hotswapref/HotswapRef.scala
Comment (catostrophe) [inline modules/core/src/main/scala/io/janstenpickle/hotswapref/HotswapRef.scala:93]: This is probably the most complex part that needs review wrt cancelation.

**Review by vasilmkd — APPROVED** [review]
I looked at the whole code, both before and after the change and I think it's quite impressive and very solid. I didn't find any assumptions with regards to cancelation that would be problematic. It follows the coding of the original Hotswap (I had to familiarize myself with it better), so this should increase the confidence in the code. I still would encourage to get another set of eyes to look over the changes, because I'm far from an expert on these matters.

File: modules/core/src/main/scala/io/janstenpickle/hotswapref/Lock.scala
Comment (catostrophe) [reply modules/core/src/main/scala/io/janstenpickle/hotswapref/Lock.scala:19]: @vasilmkd yes, for inference. This pattern is followed on the whole ce3 codebase.

**Comment by catostrophe** [comment]
@vasilmkd thanks a lot

**Comment by vasilmkd** [comment]
I may be posting too eagerly at the moment, but I might have an idea on how to streamline things a bit. I'm going to try it out in an hour when I'm back home.

File: modules/core/src/test/scala/io/janstenpickle/hotswapref/HotswapRefSpec.scala
Comment (vasilmkd) [inline modules/core/src/test/scala/io/janstenpickle/hotswapref/HotswapRefSpec.scala:51]: I think we should rework this test, it's really hard to follow. Can you please walk me through what this test aims to achieve?

File: modules/core/src/test/scala/io/janstenpickle/hotswapref/HotswapRefSpec.scala
Comment (vasilmkd) [inline modules/core/src/test/scala/io/janstenpickle/hotswapref/HotswapRefSpec.scala:55]: I can't seem to wrap my head around this sleep here. Is there a specific purpose for this?

File: modules/core/src/test/scala/io/janstenpickle/hotswapref/HotswapRefSpec.scala
Comment (vasilmkd) [inline modules/core/src/test/scala/io/janstenpickle/hotswapref/HotswapRefSpec.scala:51]: I think we should replace these sleeps with a `Deferred[IO, Unit].get` which would be completed after `update` above and after the `swap` here.

File: modules/core/src/test/scala/io/janstenpickle/hotswapref/HotswapRefSpec.scala
Comment (vasilmkd) [inline modules/core/src/test/scala/io/janstenpickle/hotswapref/HotswapRefSpec.scala:50]: The description of the test is misleading because there is no blocking on swap, otherwise the test wouldn't complete. The way `cats.effect.std.Hotswap` works is outlined in the scaladoc for the class, so the new resource is allocated _before_ the old one is released.

File: modules/core/src/test/scala/io/janstenpickle/hotswapref/HotswapRefSpec.scala
Comment (catostrophe) [reply modules/core/src/test/scala/io/janstenpickle/hotswapref/HotswapRefSpec.scala:55]: This is a leftover from a temporary version of this test, where I tracked the release of the resource. 

Yes, I agree that these tests should be reworked.

File: modules/core/src/test/scala/io/janstenpickle/hotswapref/HotswapRefSpec.scala
Comment (vasilmkd) [inline modules/core/src/test/scala/io/janstenpickle/hotswapref/HotswapRefSpec.scala:53]: Expecting this `release.get` being `false` at the end of the test somehow goes against my intuition of what `HotswapRef` tries to achieve and what is mentioned in the scaladoc. Particularly this line is confusing to me:

```scala
  * As with calls to [[swap]] on [[cats.effect.std.Hotswap]], [[swap]] will block until the previous
  * [[cats.effect.kernel.Resource]] is finalized.
```

Comment (vasilmkd) [reply modules/core/src/test/scala/io/janstenpickle/hotswapref/HotswapRefSpec.scala:53]: IMO, the scaladoc is correct and the test is not.

Comment (vasilmkd) [reply modules/core/src/test/scala/io/janstenpickle/hotswapref/HotswapRefSpec.scala:53]: This is how I would rewrite this test and how I understand it based on the `HotswapRef` scaladoc.

```scala
  it should "block on swap while handles to previous resource are held" in {
    val test = (for {
      ref0 <- ref
      ref1 <- ref
      releaseSignal <- Resource.eval(Ref.of[IO, Boolean](false))
      hotswap <- HotswapRef[IO, Ref[IO, List[String]]](Resource.pure(ref0))
      gate1 <- Resource.eval(Deferred[IO, Unit])
      _ <- Resource.eval(
        hotswap.access.use(_.update("test0" :: _) >> gate1.complete(()) >> IO.sleep(10.seconds) >> releaseSignal.set(true)).start
      )
      _ <- Resource.eval(gate1.get) // the update succeeded
      gate2 <- Resource.eval(Deferred[IO, Unit])
      _ <- Resource.eval((hotswap.swap(Resource.pure(ref1)) >> gate2.complete(())).start)
      _ <- Resource.eval(gate2.get) // the swap is complete
      _ <- Resource.eval(hotswap.access.use(_.update("test1" :: _)))
    } yield (ref0.get, ref1.get, releaseSignal.get)).use(_.tupled)

    val res = test.unsafeRunSync()

    // each ref has been used
    res._1.size should be(1)
    res._2.size should be(1)
    // access isn't blocked by hotswap
    res._1.head should be("test0")
    res._2.head should be("test1")
    // original resource isn't yet released
    res._3 should be(false) // this should be `true`
  }
```

**Comment by vasilmkd** [comment]
Here's another idea for the implementation of the `HotswapRef` which I also think satisfies the requirements. There is one lock for the whole hotswap ref (instead of a lock per resource), and the `swap` method first has to acquire all of the permits to proceed with the swapping. Of course, at this point, no other `access` can acquire a permit to continue, so we're safe there, and furthermore, it guards against other concurrent `swap` operations, because they too can't get all of the permits.

This way, there are far less allocations for each resource `swap`ped into the `HotswapRef`.

Does any of this make sense, or have I misunderstood something? Here is the code:

```scala
object HotswapRef {

  /** Creates a new [[HotswapRef]] initialized with the specified resource. The [[HotswapRef]] instance is returned
    * within a [[cats.effect.kernel.Resource]]
    */
  def apply[F[_], R](initial: Resource[F, R])(implicit F: Concurrent[F]): Resource[F, HotswapRef[F, R]] = {

    /** Secure a resource by enriching it with a semaphore-based lock and a unique token and by modifying its finalizer.
      *
      * The lock is acquired in shared mode when the resource is accessed (permits concurrent access but prohibits
      * finalization) and in exclusive mode when the resource is finalized (prohibits access).
      *
      * The token is used during access for consistent read of the holder reference.
      */
    def secure(res: Resource[F, R]): Resource[F, (R, Unique.Token)] = for {
      r <- Resource
        .makeFull((poll: Poll[F]) => poll(res.allocated)) { case (_, release) => release }
        .map(_._1)
      token <- Resource.eval(Unique[F].unique)
    } yield (r, token)

    for {
      (hotswap, securedR) <- Hotswap(secure(initial))
      holder <- Resource.eval(Ref.of(securedR))
      lock <- Resource.eval(Lock[F])
    } yield new HotswapRef[F, R] {
      override def swap(next: Resource[F, R]): F[Unit] =
        lock.exclusive.surround(hotswap.swap(secure(next).evalTap(holder.set))).void

      override val access: Resource[F, R] = {

        /** Access to the resource is protected by a shared-mode lock. The holder reference is read at least twice:
          * first, to retrieve its content, and then, after acquiring the lock, to check if the content hasn't changed
          * since the first read. If the holder has been swapped, the lock is released and the new content is passed
          * to the next step of the loop. Otherwise, it's used to build the resulting `Resource`.
          */
        val step: (R, Unique.Token) => F[Either[(R, Unique.Token), (R, F[Unit])]] =
          (r, token) =>
            F.uncancelable { poll: Poll[F] =>
              poll(lock.shared.allocated).flatMap { case (_, lockRelease) =>
                holder.get.flatMap { case (r1, token1) =>
                  if (token != token1) lockRelease.as(Left((r1, token1)))
                  else F.pure(Right((r, lockRelease)))
                }
              }
            }

        val allocated = holder.get.flatMap(_.tailRecM(step.tupled))

        Resource.makeFull((poll: Poll[F]) => poll(allocated))(_._2).map(_._1)
      }
    }
  }
}
```

**Comment by catostrophe** [comment]
@vasilmkd We use HotswapRef for holding a very frequently accessed resource, and we want `access` to be as fast as possible, i.e. to not (semantically) block if possible.

At the same time, resource allocation may take significant time (compared to access time) and may occur quite often.

In my implementation, the probability of blocking on `access` is reduced, because you only get blocked if you access the holder ref at some moment between swap and release.

**Comment by vasilmkd** [comment]
Can you please walk me through that exact situation that you described.

Calling `swap` will _not_ finish until the old resource is completely finalized, right? And this finalization waits until all `access`es to the resource are finished and have returned their permit.

So, having a single permit `Semaphore` around this ensures that only a single `swap` can _start_ (i.e. a concurrent `swap` is blocked until the previous one completely ends, including the release of the old resource).

However, a concurrent `access` is blocked until any ongoing `swap` (along with the finalization of the old resource) is ongoing.

In my proposal, a `swap` has to obtain _all_ permits to even begin. Once it begins, it can be sure that there are no ongoing `access`es because `access` and `swap` both contend on the same "global" lock. So an `access` when there is no ongoing `swap` will not block, because it only requests a single permit, which will always be available in such a situation.

I'm failing to understand the difference.

**Comment by catostrophe** [comment]
@vasilmkd in my impl:
1. While the new resource is being allocated, the ref can be accessed with no blocking 
2. When the new resource is allocated, the holder ref is set with the new resource, so subsequent calls to access aren't blocked.
3. Calls to access that happened before the ref was updated, but that couldn't get the shared lock in time, may be blocked by the exclusive lock set by release.

**Comment by catostrophe** [comment]
Calls to swap and access happen in different fibers

**Comment by vasilmkd** [comment]
Awesome, thank you for walking me through it. 😄

File: modules/core/src/main/scala/io/janstenpickle/hotswapref/HotswapRef.scala
Comment (vasilmkd) [inline modules/core/src/main/scala/io/janstenpickle/hotswapref/HotswapRef.scala:85]: Not sure if you should use `=!=` here from the `Eq[Token]` instance.

Comment (catostrophe) [reply modules/core/src/main/scala/io/janstenpickle/hotswapref/HotswapRef.scala:85]: @vasilmkd We only need to check that the ref content has changed. Firstly, I wanted to use `eq`/`ne` (referential equality) on `R`, but that would require `R <: AnyRef`, so I introduced a token. It doesn't have an `Eq` instance and `!=` on it works like `ne`.

But now I think I could just use `ne` on `Tuple` since the holder ref contains a tuple which is AnyRef.

Comment (vasilmkd) [reply modules/core/src/main/scala/io/janstenpickle/hotswapref/HotswapRef.scala:85]: Token does have an Eq instance. It's used in the CE laws.

Comment (catostrophe) [reply modules/core/src/main/scala/io/janstenpickle/hotswapref/HotswapRef.scala:85]: Hmm, I couldn't find one.

Comment (vasilmkd) [reply modules/core/src/main/scala/io/janstenpickle/hotswapref/HotswapRef.scala:85]: It should work implicitly. There's a `Hash[Token]` instance, and `Hash extends Eq`.

Comment (catostrophe) [reply modules/core/src/main/scala/io/janstenpickle/hotswapref/HotswapRef.scala:85]: Right, I've seen it. I just forgot `Hash extends Eq`.

Comment (vasilmkd) [reply modules/core/src/main/scala/io/janstenpickle/hotswapref/HotswapRef.scala:85]: Honestly, I just learned this fact today for the first time. 😆

Comment (catostrophe) [reply modules/core/src/main/scala/io/janstenpickle/hotswapref/HotswapRef.scala:85]: Does this code have any dowsides?
```scala
        val step: ((R, Lock[F])) => F[Either[(R, Lock[F]), (R, F[Unit])]] = { case tup @ (r, lock) =>
          F.uncancelable { poll: Poll[F] =>
            poll(lock.shared.allocated).flatMap { case (_, lockRelease) =>
              holder.get.flatMap { case tup1 @ (r1, lock1) =>
                if (tup ne tup1) lockRelease.as(Left((r1, lock1)))
                else F.pure(Right((r, lockRelease)))
              }
            }
          }
        }
```

Comment (vasilmkd) [reply modules/core/src/main/scala/io/janstenpickle/hotswapref/HotswapRef.scala:85]: Honestly, I feel like this code is relying on implementation details that the reference returned by `Ref` would be identical, and `Ref` never explicitly makes that guarantee. I don't know what to tell you. 😕

Comment (vasilmkd) [reply modules/core/src/main/scala/io/janstenpickle/hotswapref/HotswapRef.scala:85]: If you really want to avoid the `Unique.Token`, you can check reference equality on the `Lock` objects (I don't approve this either 😆).

Comment (catostrophe) [reply modules/core/src/main/scala/io/janstenpickle/hotswapref/HotswapRef.scala:85]: Ok, let's make it with Token and Eq.

**Comment by catostrophe** [comment]
TODO: rework tests

File: modules/core/src/main/scala/io/janstenpickle/hotswapref/HotswapRef.scala
Comment (janstenpickle) [reply modules/core/src/main/scala/io/janstenpickle/hotswapref/HotswapRef.scala]: I was going to ask the same RE `Token` and `Eq`, but looking at the default implementation of `Eq` for `Token` it uses the underlying Java hashcode and equals: 

```scala
  def fromUniversalHashCode[A]: Hash[A] =
    new Hash[A] {
      def hash(x: A) = x.hashCode()
      def eqv(x: A, y: A) = x == y
    }
```

My question is, seeing as we don't give the user the opportunity to change the implementation of `Eq`, does the indirection for the compiler introduce any runtime overhead, condsidering we are optimising `access` for frequent calls? Also, maybe less importantly, given the motivation and implementation of `Token` do we think the implementation of `Eq` is likely to change?

**Comment by janstenpickle** [comment]
Thanks for taking a look @vasilmkd, much appreciated

**Comment by janstenpickle** [comment]
> Calls to swap and access happen in different fibers

I think this is the essence of [my issue in `cats-effect`](https://github.com/typelevel/cats-effect/issues/2050). What I've not made explicit is that there's nothing stopping people from naively separating the operations of access and swap and as we've seen with this code, doing so is really difficult.

File: modules/core/src/main/scala/io/janstenpickle/hotswapref/HotswapRef.scala
Comment (vasilmkd) [reply modules/core/src/main/scala/io/janstenpickle/hotswapref/HotswapRef.scala]: `Token` is already used in other libraries for these purposes and I can guarantee you that the implementation of both `Token` and `Eq` won't change. The design of Cats Effect `Token` takes these things into consideration.

Also, `Token` is a simple `final class Token()`, whose `==` method delegates to `eq`, and the C2 compiler is able to inline all of these `eqv` and `==` calls to `eq` very easily. There is no reason to worry.

Comment (vasilmkd) [reply modules/core/src/main/scala/io/janstenpickle/hotswapref/HotswapRef.scala]: And `Token` was explicitly made to support reference equality only.

**Comment by janstenpickle** [comment]
@catostrophe the new tests look good, thanks for reworking them. I think we just need to decide on the use of `Eq` for `Token` (see my comment questioning runtime overhead above).

**Comment by catostrophe** [comment]
@vasilmkd I updated the test suite based on your recommendations and added some scenarios we discussed above.

File: modules/core/src/main/scala/io/janstenpickle/hotswapref/HotswapRef.scala
Comment (janstenpickle) [reply modules/core/src/main/scala/io/janstenpickle/hotswapref/HotswapRef.scala]: Sorry, I missed this when I made my comment below, thanks for clearing that up. I'm happy to use the `Eq` instance here then.

**Comment by vasilmkd** [comment]
Awesome,I'll take a look later today.

**Review by vasilmkd — APPROVED** [review]

**Comment by vasilmkd** [comment]
LGTM.

**Pull request merged**
